File: code-executor/scripts/typescript/parallel-execution.ts
/**
 * Script: Parallel Execution
 * Purpose: Execute multiple independent tool calls concurrently for better performance
 * Use Case: When you have multiple operations that don't depend on each other
 * MCP Tools Used: Multiple independent MCP tools (databases, APIs, services)
 *
 * How to Adapt:
 * 1. Replace example tool calls with your actual independent operations
 * 2. Adjust the number of parallel operations
 * 3. Decide on error handling strategy (fail-fast vs. partial success)
 * 4. Customize result aggregation logic
 * 5. Add timeout handling if needed
 *
 * Example Usage:
 * Fetch data from multiple sources simultaneously, aggregate analytics from
 * different services, or perform bulk operations in parallel
 *
 * Execution:
 * deno run --allow-read --allow-run --allow-env parallel-execution.ts
 */

import { callMCPTool, callMCPToolsParallel, callMCPToolsParallelSettled } from '../../lib/mcp-client.ts';

interface ParallelResult {
  success: boolean;
  completed: number;
  failed: number;
  results: any[];
  errors: string[];
  duration: number;
  speedup: string;
}

async function parallelExecution(): Promise<ParallelResult> {
  console.log('=== Starting Parallel Execution ===');
  const startTime = Date.now();

  // Define independent operations to run in parallel
  const operations = [
    {
      name: 'Fetch Users',
      call: () => callMCPTool('mcp__database__query', {
        table: 'users',
        limit: 1000
      })
    },
    {
      name: 'Fetch Orders',
      call: () => callMCPTool('mcp__database__query', {
        table: 'orders',
        limit: 1000
      })
    },
    {
      name: 'Fetch Products',
      call: () => callMCPTool('mcp__database__query', {
        table: 'products',
        limit: 500
      })
    },
    {
      name: 'Get Analytics',
      call: () => callMCPTool('mcp__analytics__getMetrics', {
        period: 'last_7_days'
      })
    },
    {
      name: 'Load Config',
      call: () => callMCPTool('mcp__config__get', {
        key: 'app_settings'
      })
    }
  ];

  console.log(`Executing ${operations.length} operations in parallel...`);

  // Execute all operations concurrently with Promise.allSettled
  // This allows us to handle partial failures gracefully
  const results = await Promise.allSettled(
    operations.map(async (op, index) => {
      const opStart = Date.now();
      console.log(`[${index + 1}/${operations.length}] Starting: ${op.name}`);

      try {
        const result = await op.call();
        const duration = Date.now() - opStart;
        console.log(`✓ [${index + 1}/${operations.length}] ${op.name} completed (${duration}ms)`);

        return {
          name: op.name,
          success: true,
          data: result,
          duration
        };
      } catch (error) {
        const duration = Date.now() - opStart;
        console.error(`✗ [${index + 1}/${operations.length}] ${op.name} failed (${duration}ms): ${error.message}`);

        return {
          name: op.name,
          success: false,
          error: error.message,
          duration
        };
      }
    })
  );

  // Analyze results
  const completed = results.filter(r => r.status === 'fulfilled' && r.value.success);
  const failed = results.filter(r => r.status === 'rejected' || (r.status === 'fulfilled' && !r.value.success));

  const successfulResults = completed.map(r => (r as PromiseFulfilledResult<any>).value);
  const errors = failed.map(r => {
    if (r.status === 'rejected') {
      return r.reason.message;
    } else {
      return (r as PromiseFulfilledResult<any>).value.error;
    }
  });

  const totalDuration = Date.now() - startTime;

  // Calculate estimated sequential time (sum of individual durations)
  const estimatedSequentialTime = successfulResults.reduce(
    (sum, r) => sum + r.duration,
    0
  );
  const speedup = estimatedSequentialTime > 0
    ? `${(estimatedSequentialTime / totalDuration).toFixed(2)}x faster`
    : 'N/A';

  console.log('\n=== Execution Summary ===');
  console.log(`Total time: ${totalDuration}ms`);
  console.log(`Estimated sequential time: ${estimatedSequentialTime}ms`);
  console.log(`Speedup: ${speedup}`);
  console.log(`Completed: ${completed.length}/${operations.length}`);
  console.log(`Failed: ${failed.length}/${operations.length}`);

  if (failed.length > 0) {
    console.warn('\nFailed operations:');
    errors.forEach((err, i) => console.warn(`  ${i + 1}. ${err}`));
  }

  return {
    success: failed.length === 0,
    completed: completed.length,
    failed: failed.length,
    results: successfulResults,
    errors,
    duration: totalDuration,
    speedup
  };
}

// Alternative: Fail-fast approach using Promise.all
// Uncomment to use this pattern instead
/*
async function parallelExecutionFailFast() {
  console.log('=== Parallel Execution (Fail-Fast) ===');

  try {
    const [users, orders, products, analytics, config] = await Promise.all([
      callMCPTool('mcp__database__query', { table: 'users' }),
      callMCPTool('mcp__database__query', { table: 'orders' }),
      callMCPTool('mcp__database__query', { table: 'products' }),
      callMCPTool('mcp__analytics__getMetrics', { period: 'last_7_days' }),
      callMCPTool('mcp__config__get', { key: 'app_settings' })
    ]);

    console.log('All operations completed successfully');

    return {
      success: true,
      data: { users, orders, products, analytics, config }
    };
  } catch (error) {
    console.error('At least one operation failed:', error.message);
    // If any operation fails, all results are lost
    throw error;
  }
}
*/

// Execute
const result = await parallelExecution();

if (result.success) {
  console.log('\n✓ All operations completed successfully');
} else {
  console.log(`\n⚠ Completed with ${result.failed} failures`);
}

File: code-executor/scripts/python/parallel_execution.py
"""
Script: Parallel Execution
Purpose: Execute multiple independent tool calls concurrently for better performance
Use Case: When you have multiple operations that don't depend on each other
MCP Tools Used: Multiple independent MCP tools (databases, APIs, services)

How to Adapt:
1. Replace example tool calls with your actual independent operations
2. Adjust the number of parallel operations
3. Decide on error handling strategy (fail-fast vs. partial success)
4. Customize result aggregation logic
5. Add timeout handling if needed

Example Usage:
Fetch data from multiple sources simultaneously, aggregate analytics from
different services, or perform bulk operations in parallel
"""

import asyncio
import json
from typing import List, Dict, Any, Callable
from datetime import datetime

async def parallel_execution() -> Dict[str, Any]:
    """Execute multiple operations in parallel with graceful error handling."""
    print('=== Starting Parallel Execution ===')
    start_time = datetime.now()

    # Define independent operations to run in parallel
    operations = [
        {
            'name': 'Fetch Users',
            'tool': 'mcp__database__query',
            'params': {'table': 'users', 'limit': 1000}
        },
        {
            'name': 'Fetch Orders',
            'tool': 'mcp__database__query',
            'params': {'table': 'orders', 'limit': 1000}
        },
        {
            'name': 'Fetch Products',
            'tool': 'mcp__database__query',
            'params': {'table': 'products', 'limit': 500}
        },
        {
            'name': 'Get Analytics',
            'tool': 'mcp__analytics__get_metrics',
            'params': {'period': 'last_7_days'}
        },
        {
            'name': 'Load Config',
            'tool': 'mcp__config__get',
            'params': {'key': 'app_settings'}
        }
    ]

    print(f'Executing {len(operations)} operations in parallel...')

    # Execute all operations concurrently with gather and return_exceptions
    # This allows us to handle partial failures gracefully
    async def execute_operation(op: Dict[str, Any], index: int) -> Dict[str, Any]:
        op_start = datetime.now()
        print(f"[{index + 1}/{len(operations)}] Starting: {op['name']}")

        try:
            result = await call_mcp_tool(op['tool'], op['params'])
            duration = (datetime.now() - op_start).total_seconds() * 1000
            print(f"✓ [{index + 1}/{len(operations)}] {op['name']} completed ({duration:.0f}ms)")

            return {
                'name': op['name'],
                'success': True,
                'data': result,
                'duration': duration
            }
        except Exception as error:
            duration = (datetime.now() - op_start).total_seconds() * 1000
            print(f"✗ [{index + 1}/{len(operations)}] {op['name']} failed ({duration:.0f}ms): {error}")

            return {
                'name': op['name'],
                'success': False,
                'error': str(error),
                'duration': duration
            }

    # Execute all operations
    results = await asyncio.gather(*[
        execute_operation(op, i) for i, op in enumerate(operations)
    ], return_exceptions=True)

    # Analyze results
    successful = [r for r in results if not isinstance(r, Exception) and r['success']]
    failed = [r for r in results if isinstance(r, Exception) or not r['success']]

    errors = []
    for r in failed:
        if isinstance(r, Exception):
            errors.append(str(r))
        else:
            errors.append(r.get('error', 'Unknown error'))

    total_duration = (datetime.now() - start_time).total_seconds() * 1000

    # Calculate estimated sequential time (sum of individual durations)
    estimated_sequential = sum(r['duration'] for r in successful)
    speedup = f"{estimated_sequential / total_duration:.2f}x faster" if estimated_sequential > 0 else 'N/A'

    print('\n=== Execution Summary ===')
    print(f'Total time: {total_duration:.0f}ms')
    print(f'Estimated sequential time: {estimated_sequential:.0f}ms')
    print(f'Speedup: {speedup}')
    print(f'Completed: {len(successful)}/{len(operations)}')
    print(f'Failed: {len(failed)}/{len(operations)}')

    if failed:
        print('\nFailed operations:')
        for i, err in enumerate(errors, 1):
            print(f'  {i}. {err}')

    return {
        'success': len(failed) == 0,
        'completed': len(successful),
        'failed': len(failed),
        'results': successful,
        'errors': errors,
        'duration': total_duration,
        'speedup': speedup
    }


# Alternative: Fail-fast approach using gather without return_exceptions
# Uncomment to use this pattern instead
"""
async def parallel_execution_fail_fast():
    print('=== Parallel Execution (Fail-Fast) ===')

    try:
        users, orders, products, analytics, config = await asyncio.gather(
            call_mcp_tool('mcp__database__query', {'table': 'users'}),
            call_mcp_tool('mcp__database__query', {'table': 'orders'}),
            call_mcp_tool('mcp__database__query', {'table': 'products'}),
            call_mcp_tool('mcp__analytics__get_metrics', {'period': 'last_7_days'}),
            call_mcp_tool('mcp__config__get', {'key': 'app_settings'})
        )

        print('All operations completed successfully')

        return {
            'success': True,
            'data': {
                'users': users,
                'orders': orders,
                'products': products,
                'analytics': analytics,
                'config': config
            }
        }
    except Exception as error:
        print(f'At least one operation failed: {error}')
        # If any operation fails, all results are lost
        raise
"""

# Execute
result = await parallel_execution()

if result['success']:
    print('\n✓ All operations completed successfully')
else:
    print(f"\n⚠ Completed with {result['failed']} failures")
